File: 2015/kt/src/test/kotlin/org/adventofcode/Day10.kt
package org.adventofcode

class Day10 {
  fun lookAndSay(s: String): String {
    return "(.)\\1*".toRegex().findAll(s).map {
      it.value.length.toString() + it.value[0]
    }.joinToString("")
  }

  fun part(s: String, n: Int): Int {
    var p = s
    for (i in 1..n) {
      p = lookAndSay(p)
    }
    return p.length
  }

  fun part1(s: String) = part(s, 40)
  fun part2(s: String) = part(s, 50)
}

File: 2015/js/day10.js
const lookAndSay = s => [...s.matchAll(/(.)\1*/g)]
  .map(v => `${v[0].length}${v[1]}`)
  .join('')

const part = (s, n) => {
  let p = s
  for (let i = 0; i < n; i++) {
    p = lookAndSay(p)
  }
  return p.length
}

const part1 = s => part(s, 40)
const part2 = s => part(s, 50)

export { lookAndSay, part1, part2 }
